src/ocean/application/Daemon.d

```diff
@@ -0,0 +1,597 @@
+
+    /// Application-level timers.
+    protected Timers timers;
+
+    /// PID lock helper.
+    private PidLock pid;
+
+    /// Signal handlers.
+    private Signals signals;
+
+    /// Process stats tracker.
+    private CpuMemoryStats system_stats;
+
+    /// TODO
+    public this ( Settings settings )
+    {
+        this.settings = settings;
+        this.files = new OpenFiles;
+        this.unix_socket = new UnixSocketCommands;
+        this.args = new Arguments;
+        this.timers = new Timers;
+        this.system_stats = new CpuMemoryStats;
+        this.initSignals();
+        this.initArgs();
+        this.initConfig();
+        this.initUnixSocketCommands();
+    }
+
+    /// TODO
+    public int run ( istring[] args )
+    {
+        scope ( exit )
+        {
+            this.unix_socket.shutdown();
+            this.pid.unlock();
+            this.signals.clear();
+        }
+
+        this.handleArgs(args);
+        this.parseConfig();
+        this.logVersion();
+        this.pid.lock();
+        initScheduler(this.settings.scheduler_config);
+        this.startTimers();
+        this.startSignalHandling();
+        this.unix_socket.startEventHandling(theScheduler.epoll);
+
+
+// TODO: possible hook / extension idea: add a protected array of delegates that
+// are all called at this point.
+
+        return TaskScheduler.runInTask(&this.mainAppLogic);
+    }
+
+    /// TODO: should include any parsing of args / config required by the app
+    protected abstract int mainAppLogic ( );
+
+    /// TODO: derived class should normally call super.initArgs()
+    protected void initArgs ( )
+    {
+        this.args("help").aliased('h').params(0)
+            .help("Display this help message and exit");
+        this.args("version").params(0)
+            .help("Show version information and exit");
+        this.args("build-info").params(0)
+            .help("Show detailed build information and exit");
+        this.args("config").aliased('c').params(1).smush()
+            .help("Use the given configuration file");
+        foreach ( conf; this.settings.config_files )
+            this.args("config").defaults(conf);
+        ConfigOverrides.setupArgs(this.args);
+    }
+
+    /// TODO
+    private void initConfig ( )
+    {
+        ConfigFiller.enable_loose_parsing(!this.settings.config_errors_fatal);
+        this.config = new ConfigParser;
+    }
+
+    /// TODO: derived class should normally call super.initArgs()
+    protected void initUnixSocketCommands ( )
+    {
+        this.unix_socket.commands.addHandler("show_version",
+            &this.showVersionCommand);
+        this.unix_socket.commands.addHandler("reopen_files",
+            &this.reopenFilesCommand);
+    }
+
+    private void initSignals ( )
+    {
+        this.signals = new Signals(&this.onSignal);
+        this.signals.handle(this.settings.signals);
+        this.signals.ignore(this.settings.ignore_signals);
+    }
```

Comment:
Some of these `init*` methods are private and some are protected.

Reply:
Yeah, there was some logic to this at some point, though it may not be consistent now. The idea was that derived classed will commonly need to override some things, whereas other things should not be overrideable.